src/connector/src/sink/remote.rs

```diff
@@ -228,14 +243,124 @@ async fn validate_remote_sink(param: &SinkParam) -> Result<()> {
+            ))
+        } else {
+            StreamChunkConverter::Other
+        }
+    }
+
+    fn convert_chunk(&self, chunk: StreamChunk) -> Result<StreamChunk> {
+        match self {
+            StreamChunkConverter::Es(es) => es.convert_chunk(chunk),
+            _ => Ok(chunk),
+        }
+    }
+}
+struct EsStreamChunkConverter {
+    json_encoder: JsonEncoder,
+    pk_indices: Vec<usize>,
+    delimiter: Option<String>,
+}
+impl EsStreamChunkConverter {
+    fn new(schema: Schema, pk_indices: Vec<usize>, delimiter: Option<String>) -> Self {
+        let json_encoder = JsonEncoder::new_with_es(schema, None);
+        Self {
+            json_encoder,
+            pk_indices,
+            delimiter,
+        }
+    }
+
+    fn convert_chunk(&self, chunk: StreamChunk) -> Result<StreamChunk> {
+        let mut ops = vec![];
+        let mut id_string_builder =
+            <Utf8ArrayBuilder as risingwave_common::array::ArrayBuilder>::new(chunk.capacity());
+        let mut json_builder =
+            <JsonbArrayBuilder as risingwave_common::array::ArrayBuilder>::new(chunk.capacity());
+        for (op, row) in chunk.rows() {
+            ops.push(op);
+            let json = JsonbVal::from(Value::Object(self.json_encoder.encode(row)?));
+            risingwave_common::array::ArrayBuilder::append(
+                &mut id_string_builder,
+                Some(&self.build_id(row)?),
+            );
+            risingwave_common::array::ArrayBuilder::append(
+                &mut json_builder,
+                Some(json.as_scalar_ref()),
+            );
+        }
+        let json_array = risingwave_common::array::ArrayBuilder::finish(json_builder);
+        let id_string_array = risingwave_common::array::ArrayBuilder::finish(id_string_builder);
+        Ok(StreamChunk::new(
+            ops,
+            vec![
+                std::sync::Arc::new(ArrayImpl::Utf8(id_string_array)),
+                std::sync::Arc::new(ArrayImpl::Jsonb(json_array)),
+            ],
+        ))
+    }
+
+    fn build_id(&self, row: RowRef<'_>) -> Result<String> {
+        if self.pk_indices.is_empty() {
+            Ok(row
+                .datum_at(0)
+                .ok_or_else(|| anyhow!("No value find in row, index is 0"))?
+                .to_text())
+        } else if self.pk_indices.len() == 1 {
+            let index = self.pk_indices.get(0).unwrap();
+            Ok(row
+                .datum_at(*index)
+                .ok_or_else(|| anyhow!("No value find in row, index is {}", index))?
+                .to_text())
+        } else {
+            let mut keys = vec![];
+            for index in &self.pk_indices {
+                keys.push(
+                    row.datum_at(*index)
+                        .ok_or_else(|| anyhow!("No value find in row, index is {}", index))?
+                        .to_text(),
+                );
+            }
+            Ok(keys.join(
+                self.delimiter
```

Comment:
We can either have a default delimiter, such as `,`, or we should find the error in validation.

A better implementation will be storing a id_builder with type `Box<dyn for Fn(row: RowRef<'_>) -> Result<String>>` when we build the stream chunk converter so that we don't have to check the pk_indices and delimiter every time we call `build_id`.